test/unit/master/test_build.py

```diff
@@ -273,6 +273,25 @@ def test_cancel_is_a_noop_if_build_is_already_finished(self):
         self.assertEqual(len(mock_slave.method_calls), num_slave_calls_before_cancel,
                          'Canceling a finished build should not cause any further calls to slave.')
 
+    def test_cancel_calls_kill_subprocesses_to_set_kill_event(self):
+        mock_slave = self._create_mock_slave()
+        build = self._create_test_build(BuildStatus.QUEUED, slaves=[mock_slave])
+        self.patch('app.master.build.util')
+        build.generate_project_type()
+
+        build.cancel()
+
+        build._project_type.kill_subprocesses.assert_called_once_with()
+
+    def test_prepare_callback_handles_atomizererror_exception_and_raises_again(self):
+        job_config = self._create_job_config()
+        subjobs = self._create_subjobs(job_config=job_config)
+        mock_subjob_calculator = self._create_mock_subjob_calc(subjobs, compute_subjob_raises_exception=True)
+        build = self._create_test_build(BuildStatus.QUEUED)
+
+        with self.assertRaises(AtomizerError):
+            build.prepare(mock_subjob_calculator)
```

Comment:
You should have a test that asserts that cancelling before calling `build.prepare()` does not compute subjobs.